src/realm/column.hpp

```diff
@@ -884,6 +885,45 @@ int ColumnBase::compare_values(const Column* column, size_t row1, size_t row2) n
     return a == b ? 0 : a < b ? 1 : -1;
 }
 
+namespace _impl {
+template <int> struct IntTypeForSize;
+template <> struct IntTypeForSize<1> { using type = uint8_t; };
+template <> struct IntTypeForSize<2> { using type = uint16_t; };
+template <> struct IntTypeForSize<4> { using type = uint32_t; };
+template <> struct IntTypeForSize<8> { using type = uint64_t; };
+
+template <typename Float>
+int compare_float(Float a_raw, Float b_raw)
+{
+    // nans (and by extension nulls) are treated as being less than all non-nan values
+    bool a_nan = std::isnan(a_raw);
+    bool b_nan = std::isnan(b_raw);
+    if (a_nan != b_nan) {
+        return a_nan ? 1 : -1;
+    }
+
+    // Compare the values as ints, which gives correct results for IEEE floats
+    // and bypasses the usual behavior of nans not being comparable to each other
+    using IntType = typename _impl::IntTypeForSize<sizeof(Float)>::type;
+    IntType a = 0, b = 0;
+    memcpy(&a, &a_raw, sizeof(Float));
```

Comment:
Is a copy required? Could reinterpret_cast not have solved the problem?

Reply:
reinterpret_cast here is an aliasing violation. The copy is optimized out in release builds.

Reply:
OK - fine